In [215]:
import pandas as pd

# MAVEN ANALYTICS PIZZA CHALLENEGE

In [216]:
'''Here are some questions that we'd like to be able to answer:

What days and times do we tend to be busiest?
How many pizzas are we making during peak periods?
What are our best and worst selling pizzas?
What's our average order value?
How well are we utilizing our seating capacity? (we have 15 tables and 60 seats)'''

"Here are some questions that we'd like to be able to answer:\n\nWhat days and times do we tend to be busiest?\nHow many pizzas are we making during peak periods?\nWhat are our best and worst selling pizzas?\nWhat's our average order value?\nHow well are we utilizing our seating capacity? (we have 15 tables and 60 seats)"

In [217]:
df1=pd.read_csv('order_details.csv')

In [218]:
df2=pd.read_csv('orders.csv')

In [219]:
df4=pd.read_csv('pizzas.csv')

In [220]:
df3=pd.read_csv('pizza_types.csv',encoding='latin1')#there were some special characters in pizza_types so we are using this encoding to avoid error

In [221]:
df1.head()

order_details_id  order_id       pizza_id  quantity
0                 1         1     hawaiian_m         1
1                 2         2  classic_dlx_m         1
2                 3         2  five_cheese_l         1
3                 4         2    ital_supr_l         1
4                 5         2     mexicana_m         1

In [222]:
df2.head()

order_id        date      time
0         1  2015-01-01  11:38:36
1         2  2015-01-01  11:57:40
2         3  2015-01-01  12:12:28
3         4  2015-01-01  12:16:31
4         5  2015-01-01  12:21:30

In [223]:
df3.head()

pizza_type_id                          name category  \
0       bbq_ckn    The Barbecue Chicken Pizza  Chicken   
1      cali_ckn  The California Chicken Pizza  Chicken   
2   ckn_alfredo     The Chicken Alfredo Pizza  Chicken   
3     ckn_pesto       The Chicken Pesto Pizza  Chicken   
4    southw_ckn   The Southwest Chicken Pizza  Chicken   

                                         ingredients  
0  Barbecued Chicken, Red Peppers, Green Peppers,...  
1  Chicken, Artichoke, Spinach, Garlic, Jalapeno ...  
2  Chicken, Red Onions, Red Peppers, Mushrooms, A...  
3  Chicken, Tomatoes, Red Peppers, Spinach, Garli...  
4  Chicken, Tomatoes, Red Peppers, Red Onions, Ja...

In [224]:
df4.head()

pizza_id pizza_type_id size  price
0   bbq_ckn_s       bbq_ckn    S  12.75
1   bbq_ckn_m       bbq_ckn    M  16.75
2   bbq_ckn_l       bbq_ckn    L  20.75
3  cali_ckn_s      cali_ckn    S  12.75
4  cali_ckn_m      cali_ckn    M  16.75

In [225]:
df1.drop('order_details_id',axis=1,inplace=True)#removing unwanted column

In [226]:
df1.head()

order_id       pizza_id  quantity
0         1     hawaiian_m         1
1         2  classic_dlx_m         1
2         2  five_cheese_l         1
3         2    ital_supr_l         1
4         2     mexicana_m         1

In [227]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48620 entries, 0 to 48619
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   order_id  48620 non-null  int64 
 1   pizza_id  48620 non-null  object
 2   quantity  48620 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 1.1+ MB


In [228]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21350 entries, 0 to 21349
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   order_id  21350 non-null  int64 
 1   date      21350 non-null  object
 2   time      21350 non-null  object
dtypes: int64(1), object(2)
memory usage: 500.5+ KB


In [229]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   pizza_type_id  32 non-null     object
 1   name           32 non-null     object
 2   category       32 non-null     object
 3   ingredients    32 non-null     object
dtypes: object(4)
memory usage: 1.1+ KB


In [230]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96 entries, 0 to 95
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   pizza_id       96 non-null     object 
 1   pizza_type_id  96 non-null     object 
 2   size           96 non-null     object 
 3   price          96 non-null     float64
dtypes: float64(1), object(3)
memory usage: 3.1+ KB


In [231]:
df1=pd.merge(df1,df2,on='order_id')

In [232]:
df1.head()

order_id       pizza_id  quantity        date      time
0         1     hawaiian_m         1  2015-01-01  11:38:36
1         2  classic_dlx_m         1  2015-01-01  11:57:40
2         2  five_cheese_l         1  2015-01-01  11:57:40
3         2    ital_supr_l         1  2015-01-01  11:57:40
4         2     mexicana_m         1  2015-01-01  11:57:40

In [233]:
df1['pizza_id'].unique()

array(['hawaiian_m', 'classic_dlx_m', 'five_cheese_l', 'ital_supr_l',
       'mexicana_m', 'thai_ckn_l', 'ital_supr_m', 'prsc_argla_l',
       'bbq_ckn_s', 'the_greek_s', 'spinach_supr_s', 'classic_dlx_s',
       'green_garden_s', 'ital_cpcllo_l', 'ital_supr_s', 'mexicana_s',
       'spicy_ital_l', 'spin_pesto_l', 'veggie_veg_s', 'mexicana_l',
       'southw_ckn_l', 'bbq_ckn_l', 'cali_ckn_l', 'cali_ckn_m',
       'pepperoni_l', 'cali_ckn_s', 'ckn_pesto_l', 'big_meat_s',
       'soppressata_l', 'four_cheese_l', 'napolitana_s', 'calabrese_m',
       'four_cheese_m', 'ital_veggie_s', 'mediterraneo_m',
       'peppr_salami_s', 'spinach_fet_l', 'napolitana_l', 'sicilian_l',
       'ital_cpcllo_m', 'southw_ckn_s', 'bbq_ckn_m', 'pepperoni_m',
       'prsc_argla_s', 'sicilian_m', 'veggie_veg_l', 'ckn_alfredo_s',
       'pepperoni_s', 'green_garden_l', 'green_garden_m', 'pep_msh_pep_l',
       'hawaiian_s', 'peppr_salami_m', 'ckn_alfredo_m', 'peppr_salami_l',
       'spin_pesto_s', 'thai_ckn_m'

In [234]:
df4['pizza_id'].unique()

array(['bbq_ckn_s', 'bbq_ckn_m', 'bbq_ckn_l', 'cali_ckn_s', 'cali_ckn_m',
       'cali_ckn_l', 'ckn_alfredo_s', 'ckn_alfredo_m', 'ckn_alfredo_l',
       'ckn_pesto_s', 'ckn_pesto_m', 'ckn_pesto_l', 'southw_ckn_s',
       'southw_ckn_m', 'southw_ckn_l', 'thai_ckn_s', 'thai_ckn_m',
       'thai_ckn_l', 'big_meat_s', 'big_meat_m', 'big_meat_l',
       'classic_dlx_s', 'classic_dlx_m', 'classic_dlx_l', 'hawaiian_s',
       'hawaiian_m', 'hawaiian_l', 'ital_cpcllo_s', 'ital_cpcllo_m',
       'ital_cpcllo_l', 'napolitana_s', 'napolitana_m', 'napolitana_l',
       'pep_msh_pep_s', 'pep_msh_pep_m', 'pep_msh_pep_l', 'pepperoni_s',
       'pepperoni_m', 'pepperoni_l', 'the_greek_s', 'the_greek_m',
       'the_greek_l', 'the_greek_xl', 'the_greek_xxl', 'brie_carre_s',
       'calabrese_s', 'calabrese_m', 'calabrese_l', 'ital_supr_s',
       'ital_supr_m', 'ital_supr_l', 'peppr_salami_s', 'peppr_salami_m',
       'peppr_salami_l', 'prsc_argla_s', 'prsc_argla_m', 'prsc_argla_l',
       'sicilian_s'

In [235]:
df1=pd.merge(df1,df4,on='pizza_id')

In [236]:
df1.head()

order_id    pizza_id  quantity        date      time pizza_type_id size  \
0         1  hawaiian_m         1  2015-01-01  11:38:36      hawaiian    M   
1        77  hawaiian_m         1  2015-01-02  12:22:46      hawaiian    M   
2       146  hawaiian_m         1  2015-01-03  14:22:10      hawaiian    M   
3       163  hawaiian_m         1  2015-01-03  16:54:54      hawaiian    M   
4       247  hawaiian_m         1  2015-01-04  20:55:29      hawaiian    M   

   price  
0  13.25  
1  13.25  
2  13.25  
3  13.25  
4  13.25

In [237]:
df1=pd.merge(df1,df3,on='pizza_type_id')

In [238]:
df1.head()

order_id    pizza_id  quantity        date      time pizza_type_id size  \
0         1  hawaiian_m         1  2015-01-01  11:38:36      hawaiian    M   
1        77  hawaiian_m         1  2015-01-02  12:22:46      hawaiian    M   
2       146  hawaiian_m         1  2015-01-03  14:22:10      hawaiian    M   
3       163  hawaiian_m         1  2015-01-03  16:54:54      hawaiian    M   
4       247  hawaiian_m         1  2015-01-04  20:55:29      hawaiian    M   

   price                name category  \
0  13.25  The Hawaiian Pizza  Classic   
1  13.25  The Hawaiian Pizza  Classic   
2  13.25  The Hawaiian Pizza  Classic   
3  13.25  The Hawaiian Pizza  Classic   
4  13.25  The Hawaiian Pizza  Classic   

                                ingredients  
0  Sliced Ham, Pineapple, Mozzarella Cheese  
1  Sliced Ham, Pineapple, Mozzarella Cheese  
2  Sliced Ham, Pineapple, Mozzarella Cheese  
3  Sliced Ham, Pineapple, Mozzarella Cheese  
4  Sliced Ham, Pineapple, Mozzarella Cheese

In [240]:
df1.shape

(48620, 11)

In [241]:
df1.to_excel('Maven Analytics.xlsx',index=False)

In [243]:
df_best_worst_pizza=df1.groupby('name')['quantity'].sum()

What are our best and worst selling pizzas?

In [250]:
df_best_worst_pizza.sort_values()

name
The Brie Carre Pizza                           490
The Mediterranean Pizza                        934
The Calabrese Pizza                            937
The Spinach Supreme Pizza                      950
The Soppressata Pizza                          961
The Spinach Pesto Pizza                        970
The Chicken Pesto Pizza                        973
The Italian Vegetables Pizza                   981
The Chicken Alfredo Pizza                      987
The Green Garden Pizza                         997
The Pepperoni, Mushroom, and Peppers Pizza    1359
The Five Cheese Pizza                         1409
The Greek Pizza                               1420
The Italian Capocollo Pizza                   1438
The Pepper Salami Pizza                       1446
The Spinach and Feta Pizza                    1446
The Prosciutto and Arugula Pizza              1457
The Napolitana Pizza                          1464
The Mexicana Pizza                            1484
The Vegetables + Vegetable

Worst performing pizza is The Brie Carre Pizza 
Best performing pizza is The Classic Deluxe Pizza   


What's our average order value?

In [251]:
df1.sort_values(by='order_id')

order_id       pizza_id  quantity        date      time pizza_type_id  \
0             1     hawaiian_m         1  2015-01-01  11:38:36      hawaiian   
9450          2     thai_ckn_l         1  2015-01-01  11:57:40      thai_ckn   
7994          2     mexicana_m         1  2015-01-01  11:57:40      mexicana   
2370          2  classic_dlx_m         1  2015-01-01  11:57:40   classic_dlx   
6145          2    ital_supr_l         1  2015-01-01  11:57:40     ital_supr   
...         ...            ...       ...         ...       ...           ...   
37268     21348   napolitana_s         1  2015-12-31  21:23:10    napolitana   
46610     21348  ckn_alfredo_m         1  2015-12-31  21:23:10   ckn_alfredo   
36223     21348  four_cheese_l         1  2015-12-31  21:23:10   four_cheese   
9449      21349     mexicana_l         1  2015-12-31  22:09:54      mexicana   
13671     21350      bbq_ckn_s         1  2015-12-31  23:02:05       bbq_ckn   

      size  price                        name category  \
0        M  13.25          The Hawaiian Pizza  Classic   
9450     L  20.75      The Thai Chicken Pizza  Chicken   
7994     M  16.00          The Mexicana Pizza   Veggie   
2370     M  16.00    The Classic Deluxe Pizza  Classic   
6145     L  20.75   The Italian Supreme Pizza  Supreme   
...    ...    ...                         ...      ...   
37268    S  12.00        The Napolitana Pizza  Classic   
46610    M  16.75   The Chicken Alfredo Pizza  Chicken   
36223    L  17.95       The Four Cheese Pizza   Veggie   
9449     L  20.25          The Mexicana Pizza   Veggie   
13671    S  12.75  The Barbecue Chicken Pizza  Chicken   

                                             ingredients  
0               Sliced Ham, Pineapple, Mozzarella Cheese  
9450   Chicken, Pineapple, Tomatoes, Red Peppers, Tha...  
7994   Tomatoes, Red Peppers, Jalapeno Peppers, Red O...  
2370   Pepperoni, Mushrooms, Red Onions, Red Peppers,...  
6145   Calabrese Salami, Capocollo, Tomatoes, Red Oni...  
...                                                  ...  
37268  Tomatoes, Anchovies, Green Olives, Red Onions,...  
46610  Chicken, Red Onions, Red Peppers, Mushrooms, A...  
36223  Ricotta Cheese, Gorgonzola Piccante Cheese, Mo...  
9449   Tomatoes, Red Peppers, Jalapeno Peppers, Red O...  
13671  Barbecued Chicken, Red Peppers, Green Peppers,...  

[48620 rows x 11 columns]

In [253]:
df1['order_part_total']=df1['quantity']*df1['price']

#In df1 order_part_total is used to calculate the total amount spent on one kind of pizza of an order ,that is, we could order more than one kinf of pizza with the same order number.How much are we spending on each kind of pizza in our order is seperate,eg:we could have order number 2 in which have ordered thain_ckn and spent 20.75 and mexicana and spent 16.00,this 20.75 is price * quantity i.e=20.75.Later on we add 20.75+16.00 to say total spent on order is 36.75.This way we get the total amount spent for each order under 'order_part_total' in df_order_value_avaerage.From there wecan get average value per order.

In [254]:
df1.head()

order_id    pizza_id  quantity        date      time pizza_type_id size  \
0         1  hawaiian_m         1  2015-01-01  11:38:36      hawaiian    M   
1        77  hawaiian_m         1  2015-01-02  12:22:46      hawaiian    M   
2       146  hawaiian_m         1  2015-01-03  14:22:10      hawaiian    M   
3       163  hawaiian_m         1  2015-01-03  16:54:54      hawaiian    M   
4       247  hawaiian_m         1  2015-01-04  20:55:29      hawaiian    M   

   price                name category  \
0  13.25  The Hawaiian Pizza  Classic   
1  13.25  The Hawaiian Pizza  Classic   
2  13.25  The Hawaiian Pizza  Classic   
3  13.25  The Hawaiian Pizza  Classic   
4  13.25  The Hawaiian Pizza  Classic   

                                ingredients  order_part_total  
0  Sliced Ham, Pineapple, Mozzarella Cheese             13.25  
1  Sliced Ham, Pineapple, Mozzarella Cheese             13.25  
2  Sliced Ham, Pineapple, Mozzarella Cheese             13.25  
3  Sliced Ham, Pineapple, Mozzarella Cheese             13.25  
4  Sliced Ham, Pineapple, Mozzarella Cheese             13.25

In [263]:
df_order_value_average=df1.groupby('order_id')['order_part_total'].sum()

In [264]:
df_order_value_average

order_id  order_part_total
0             1             13.25
1            77             13.25
2           146             13.25
3           163             13.25
4           247             13.25
...         ...               ...
48615     21167             23.65
48616     21198             23.65
48617     21225             23.65
48618     21282             23.65
48619     21308             23.65

[48620 rows x 2 columns]

In [267]:
df_order_value_average=pd.DataFrame(df_order_value_average)

In [269]:
df_order_value_average.columns

Index(['order_id', 'order_part_total'], dtype='object')

In [270]:
df_order_value_average['order_part_total'].mean()

16.82147367338452

average order value is 16.82147367338452

categorising time into morning,afternoon,evening and night

In [311]:
df1['hour']=pd.to_datetime(df1['time']).dt.hour

In [312]:
df1.head()

order_id    pizza_id  quantity        date      time pizza_type_id size  \
0         1  hawaiian_m         1  2015-01-01  11:38:36      hawaiian    M   
1        77  hawaiian_m         1  2015-01-02  12:22:46      hawaiian    M   
2       146  hawaiian_m         1  2015-01-03  14:22:10      hawaiian    M   
3       163  hawaiian_m         1  2015-01-03  16:54:54      hawaiian    M   
4       247  hawaiian_m         1  2015-01-04  20:55:29      hawaiian    M   

   price                name category  \
0  13.25  The Hawaiian Pizza  Classic   
1  13.25  The Hawaiian Pizza  Classic   
2  13.25  The Hawaiian Pizza  Classic   
3  13.25  The Hawaiian Pizza  Classic   
4  13.25  The Hawaiian Pizza  Classic   

                                ingredients  order_part_total  hour  
0  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    11  
1  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    12  
2  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    14  
3  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    16  
4  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    20

In [313]:
df1.hour.unique()

array([11, 12, 14, 16, 20, 17, 13, 18, 19, 15, 21, 22, 23,  9, 10])

In [314]:
def time_category(i):
    
        if i>=9 and i<12:
            return 'morning'
        elif i>=12 and i<17:
            return 'afternoon'
        elif i>=17 and i<=19:
            return 'evening'
        else:
            return 'night'


In [315]:
df1['time_of_day']=df1['hour'].apply(time_category)

In [317]:
df1.dtypes

order_id              int64
pizza_id             object
quantity              int64
date                 object
time                 object
pizza_type_id        object
size                 object
price               float64
name                 object
category             object
ingredients          object
order_part_total    float64
hour                  int64
time_of_day          object
dtype: object

In [318]:
df1.head()

order_id    pizza_id  quantity        date      time pizza_type_id size  \
0         1  hawaiian_m         1  2015-01-01  11:38:36      hawaiian    M   
1        77  hawaiian_m         1  2015-01-02  12:22:46      hawaiian    M   
2       146  hawaiian_m         1  2015-01-03  14:22:10      hawaiian    M   
3       163  hawaiian_m         1  2015-01-03  16:54:54      hawaiian    M   
4       247  hawaiian_m         1  2015-01-04  20:55:29      hawaiian    M   

   price                name category  \
0  13.25  The Hawaiian Pizza  Classic   
1  13.25  The Hawaiian Pizza  Classic   
2  13.25  The Hawaiian Pizza  Classic   
3  13.25  The Hawaiian Pizza  Classic   
4  13.25  The Hawaiian Pizza  Classic   

                                ingredients  order_part_total  hour  \
0  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    11   
1  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    12   
2  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    14   
3  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    16   
4  Sliced Ham, Pineapple, Mozzarella Cheese             13.25    20   

  time_of_day  
0     morning  
1   afternoon  
2   afternoon  
3   afternoon  
4       night

In [319]:
#extracting day from from date

In [329]:
df1['date']=pd.to_datetime(df1['date'])#converting from object to date and time

In [330]:
df1.dtypes

order_id                     int64
pizza_id                    object
quantity                     int64
date                datetime64[ns]
time                        object
pizza_type_id               object
size                        object
price                      float64
name                        object
category                    object
ingredients                 object
order_part_total           float64
hour                         int64
time_of_day                 object
day                 datetime64[ns]
dtype: object

In [339]:
df1['day']=df1['date'].dt.dayofweek

In [340]:
df1['day']

0        3
1        4
2        5
3        5
4        6
        ..
48615    6
48616    0
48617    1
48618    3
48619    3
Name: day, Length: 48620, dtype: int64

In [342]:
day_in_words={0:'Monday',1:'Tuesday',2:'Wednersday',3:'Thursday',4:'Friday',5:'Saturday',6:'Sunday'}

In [343]:
df1['day']=df1['day'].map(day_in_words)

In [344]:
df1['day']

0        Thursday
1          Friday
2        Saturday
3        Saturday
4          Sunday
           ...   
48615      Sunday
48616      Monday
48617     Tuesday
48618    Thursday
48619    Thursday
Name: day, Length: 48620, dtype: object

In [ ]:
What days and times do we tend to be busiest?

In [341]:
#busiest days of the week

array([3, 4, 5, 6, 1, 2, 0])

In [352]:
df1.groupby('day')['order_id'].nunique().sort_values()

day
Sunday        2624
Monday        2794
Tuesday       2973
Wednersday    3024
Saturday      3158
Thursday      3239
Friday        3538
Name: order_id, dtype: int64

Saturday,Friday,Thursday seem to be far more busier than other days in terms of number of orders

In [353]:
#busiest time of the day

In [356]:
df1.groupby(['day','time_of_day'])['order_id'].nunique().sort_values()

day         time_of_day
Saturday    morning          86
Sunday      morning          96
Tuesday     morning         189
Friday      morning         192
Thursday    morning         217
Monday      morning         227
Wednersday  morning         233
Monday      night           362
Thursday    night           393
Wednersday  night           404
Sunday      night           416
Tuesday     night           426
Saturday    night           752
Friday      night           778
Monday      evening         855
Tuesday     evening         884
Wednersday  evening         916
Sunday      evening         927
Thursday    evening        1007
Friday      evening        1071
Saturday    evening        1084
Sunday      afternoon      1185
Saturday    afternoon      1236
Monday      afternoon      1350
Wednersday  afternoon      1471
Tuesday     afternoon      1474
Friday      afternoon      1497
Thursday    afternoon      1622
Name: order_id, dtype: int64

In [357]:
#afternoon seems to be the busiest time of the day from 12 to 5

How many pizzas are we making during peak periods?

In [360]:
df_peak_hours=df1[(df1['time_of_day']=='afternoon') | (df1['time_of_day']=='evening')]

In [364]:
df_peak_hours.groupby(['day','time_of_day'])['quantity'].sum().sort_values()

day         time_of_day
Tuesday     evening        1908
Monday      evening        1915
Sunday      evening        2023
Wednersday  evening        2044
Thursday    evening        2250
Friday      evening        2443
Saturday    evening        2451
Sunday      afternoon      2922
Saturday    afternoon      3128
Monday      afternoon      3341
Wednersday  afternoon      3548
Tuesday     afternoon      3661
Friday      afternoon      3685
Thursday    afternoon      3972
Name: quantity, dtype: int64

How well are we utilizing our seating capacity? (we have 15 tables and 60 seats)

In [370]:
#not enough information available to explore this concept